### PR TITLE
Fix #1269: Migrate to Spring Boot 3

### LINKED ISSUE #1269: Migrate to Spring Boot 3
Analyze migration to Spring Boot 3 and upgrade all required project dependencies.

### PR DESCRIPTION
**Pull request by romanstrobl**
Web Flow has been migrated to Spring Authorization Server and Spring Security 6 including migration of OAuth 2.1 resource server and OAuth 2.1 client.

Pending tasks:
- [x] Migrate TPP engine to new OAuth 2.1 repositories and services.
- [x] Review and update documentation related to OAuth 2.x and database changes.
- [x] Revert unwated changes in existing Liquibase migration scripts.
- [x] Prepare Liquibase migration scripts.
- [x] Review configuration properties, unify their naming conventions.
- [x] Consider what are reasonable default time-to-live values for access tokens and refresh tokens.
- [x] Add more details about client upgrade to this version into the migration guide.
- [x] Prepare default configuration of OAuth 2.1 client in sample initial data.
- [x] Add comment about Java 17 into migration guide.
- [x] Add comment about revocation of access tokens after migration to Spring authorization server.
- [x] Improve error handling in case of invalid authorization requests.
- [x] Review all changes.

### CODE REVIEW

File: docs/db/changelog/changesets/powerauth-nextstep/1.4.x/20230324-init-db.xml [hunk 1]
```diff
@@ -781,10 +781,10 @@
                 <constraints nullable="false" />
             </column>
             <column name="operation_form_data" type="text" remarks="Structured data of the operation that are displayed to the end user." />
-            <column name="application_id" type="varchar(256)" remarks="ID of the application that initiated the operation, usually OAuth 2.0 client ID." />
+            <column name="application_id" type="varchar(256)" remarks="ID of the application that initiated the operation, usually OAuth 2.1 client ID." />
```
Comment (banterCZ) [inline docs/db/changelog/changesets/powerauth-nextstep/1.4.x/20230324-init-db.xml:784]: Doesn't it affect the checksum?

File: powerauth-nextstep/src/main/java/io/getlime/security/powerauth/app/nextstep/configuration/SecurityConfig.java [hunk 2]
```diff
@@ -17,27 +17,30 @@
...
+import org.springframework.security.web.SecurityFilterChain;
 
 /**
  * Default Spring Security configuration.
  *
  * @author Petr Dvorak, petr@wultra.com
  */
 @Configuration
-public class SecurityConfig extends WebSecurityConfigurerAdapter {
+public class SecurityConfig {
```
Comment (banterCZ) [inline powerauth-nextstep/src/main/java/io/getlime/security/powerauth/app/nextstep/configuration/SecurityConfig.java:31]: I believe that we should add `@EnableWebSecurity`.

File: powerauth-tpp-engine/src/main/java/io/getlime/security/powerauth/app/tppengine/configuration/SecurityConfig.java [hunk 3]
```diff
@@ -18,27 +18,29 @@
...
+import org.springframework.security.web.SecurityFilterChain;
 
 /**
  * Default Spring Security configuration.
  *
  * @author Petr Dvorak, petr@wultra.com
  */
 @Configuration
-public class SecurityConfig extends WebSecurityConfigurerAdapter {
+public class SecurityConfig {
```
Comment (banterCZ) [inline powerauth-tpp-engine/src/main/java/io/getlime/security/powerauth/app/tppengine/configuration/SecurityConfig.java:32]: I believe that we should add `@EnableWebSecurity`.

File: powerauth-webflow-authentication/pom.xml [hunk 4]
```diff
@@ -115,6 +115,12 @@
             <artifactId>spring-boot-starter-test</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>javax.xml.bind</groupId>
```
Comment (banterCZ) [inline powerauth-webflow-authentication/pom.xml:119]: 🤔 Do we really need the`javax` dependency? It was also migrated to `jakarta`.
```xml
<groupId>jakarta.xml.bind</groupId>
<artifactId>jakarta.xml.bind-api</artifactId>
<version>4.0.0</version>
```

File: powerauth-webflow-client/pom.xml [hunk 5]
```diff
@@ -92,8 +74,16 @@
...
-            <artifactId>httpclient</artifactId>
+            <groupId>org.apache.httpcomponents.client5</groupId>
+            <artifactId>httpclient5</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-webmvc</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.h2database</groupId>
```
Comment (banterCZ) [inline powerauth-webflow-client/pom.xml:85]: Do we really need `h2` dependency in the _compile_ scope?

File: powerauth-webflow-client/src/main/java/io/getlime/security/powerauth/app/webflow/demo/configuration/SecurityConfiguration.java [hunk 6]
```diff
@@ -18,20 +18,106 @@
...
+                        .requestMatchers("/", "/home", "/css/**", "/js/**", "/images/**").permitAll()
+                        .anyRequest().fullyAuthenticated()
+                )
+                .oauth2Login()
+                    .authorizedClientRepository(authorizedClientRepository)
+                    .authorizedClientService(authorizedClientService)
+                    .clientRegistrationRepository(clientRegistrationRepository)
+                    .authorizationEndpoint()
+                    .authorizationRequestResolver(new CustomAuthorizationRequestResolver(this.clientRegistrationRepository))
+                .and()
```
Comment (banterCZ) [inline powerauth-webflow-client/src/main/java/io/getlime/security/powerauth/app/webflow/demo/configuration/SecurityConfiguration.java:74]: It is possible to get rid of `and()` but we may postpone to a different pull request https://spring.io/blog/2019/11/21/spring-security-lambda-dsl

File: powerauth-webflow-resources/src/main/java/io/getlime/security/powerauth/lib/webflow/resource/controller/UserProfileController.java [hunk 7]
```diff
@@ -152,7 +151,7 @@ public UserProfileController(DataAdapterClient client, AuthorizationServerTokenS
             logger.info("Found user with ID: {}, given name: {}, family name: {}", usedId, givenName, familyName);
             return new UserInfoResponse(id, id, givenName, familyName, extras);
         } catch (DataAdapterClientErrorException e) {
-            throw new UnauthorizedUserException("Unable to fetch user details from data adapter");
+            throw new OAuth2AuthenticationException("Unable to fetch user details from data adapter");
```
Comment (banterCZ) [inline powerauth-webflow-resources/src/main/java/io/getlime/security/powerauth/lib/webflow/resource/controller/UserProfileController.java:154]: Let's propagate the cause exception or log it at least.

File: docs/db/changelog/changesets/powerauth-nextstep/1.4.x/20230324-init-db.xml (hunk 1 above)
Comment (romanstrobl) [reply docs/db/changelog/changesets/powerauth-nextstep/1.4.x/20230324-init-db.xml]: Yes indeed, I reverted these changes.

File: powerauth-nextstep/src/main/java/io/getlime/security/powerauth/app/nextstep/configuration/SecurityConfig.java (hunk 2 above)
Comment (romanstrobl) [reply powerauth-nextstep/src/main/java/io/getlime/security/powerauth/app/nextstep/configuration/SecurityConfig.java:33]: Sure, added.

File: powerauth-tpp-engine/src/main/java/io/getlime/security/powerauth/app/tppengine/configuration/SecurityConfig.java (hunk 3 above)
Comment (romanstrobl) [reply powerauth-tpp-engine/src/main/java/io/getlime/security/powerauth/app/tppengine/configuration/SecurityConfig.java:34]: Sure, added.

File: powerauth-webflow-authentication/pom.xml (hunk 4 above)
Comment (romanstrobl) [reply powerauth-webflow-authentication/pom.xml:119]: I will remove this dependency, it should not be needed anymore.

File: powerauth-webflow-client/pom.xml (hunk 5 above)
Comment (romanstrobl) [reply powerauth-webflow-client/pom.xml:85]: We can remove `h2` dependency completely, the client should use in memory repositories after migration from `spring-social` to `spring-security-oauth2-client`.

File: docs/db/changelog/changesets/powerauth-webflow/1.5.x/20230510-spring-authorization-server-1.5.0.xml [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+    </changeSet>
+
+    <!-- Migration to Spring authorization server: drop table oauth_client_details -->
+    <changeSet id="4" logicalFilePath="powerauth-webflow/1.5.x/20230510-spring-authorization-server-1.5.0.xml" author="Roman Strobl">
+        <preConditions onFail="MARK_RAN">
+            <tableExists tableName="oauth_client_details"/>
+        </preConditions>
+        <comment>Drop table oauth_client_details</comment>
+        <dropTable tableName="oauth_client_details"/>
+        <!-- no rollback -->
```
Comment (banterCZ) [inline docs/db/changelog/changesets/powerauth-webflow/1.5.x/20230510-spring-authorization-server-1.5.0.xml:136]: 🤔 We do not work much with rollbacks yet, but to be perfect, rollback for `drop` could be create table from the init script. But is a quite philosophical question.

File: powerauth-webflow/src/main/java/io/getlime/security/powerauth/app/webflow/configuration/SecurityConfiguration.java [hunk 9]
```diff
@@ -19,67 +19,131 @@
...
+import java.util.Objects;
 
 /**
  * Default Spring Security configuration.
  *
  * @author Petr Dvorak, petr@wultra.com
  */
 @Configuration
-public class SecurityConfiguration extends WebSecurityConfigurerAdapter {
+public class SecurityConfiguration {
```
Comment (banterCZ) [inline powerauth-webflow/src/main/java/io/getlime/security/powerauth/app/webflow/configuration/SecurityConfiguration.java:58]:
```suggestion
@EnableWebSecurity
public class SecurityConfiguration {
```

File: powerauth-webflow/src/main/java/io/getlime/security/powerauth/app/webflow/configuration/SecurityConfiguration.java [hunk 10]
```diff
@@ -19,67 +19,131 @@
...
+                        .opaqueToken((opaque) -> opaque
+                                .introspectionUri(this.introspectionUri)
+                                .introspectionClientCredentials(this.clientId, this.clientSecret)
+                        )
+                )
+                // Configure securityContextRepository for session managemet, see: https://docs.spring.io/spring-security/reference/migration/servlet/session-management.html
+                .securityContext((securityContext) -> securityContext
+                        .securityContextRepository(securityContextRepository)
+                )
+                .csrf()
```
Comment (banterCZ) [inline powerauth-webflow/src/main/java/io/getlime/security/powerauth/app/webflow/configuration/SecurityConfiguration.java:115]: It is possible to get rid of and() but we may postpone to a different pull request https://spring.io/blog/2019/11/21/spring-security-lambda-dsl

**Review by banterCZ — APPROVED** [review]
Great job!

File: docs/db/changelog/changesets/powerauth-webflow/1.5.x/20230510-spring-authorization-server-1.5.0.xml [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+            <column name="user_code_value" type="text"/>
+            <column name="user_code_issued_at" type="timestamp"/>
+            <column name="user_code_expires_at" type="timestamp"/>
+            <column name="user_code_metadata" type="text"/>
+            <column name="device_code_value" type="text"/>
+            <column name="device_code_issued_at" type="timestamp"/>
+            <column name="device_code_expires_at" type="timestamp"/>
+            <column name="device_code_metadata" type="text"/>
+        </createTable>
+        <!-- no rollback -->
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply docs/db/changelog/changesets/powerauth-webflow/1.5.x/20230510-spring-authorization-server-1.5.0.xml:57]: I can prepare the rollback script as well.

File: docs/db/changelog/changesets/powerauth-webflow/1.5.x/20230510-spring-authorization-server-1.5.0.xml (hunk 8 above)
Comment (romanstrobl) [reply docs/db/changelog/changesets/powerauth-webflow/1.5.x/20230510-spring-authorization-server-1.5.0.xml:136]: Yes, I can prepare the rollback script.

File: powerauth-webflow/src/main/java/io/getlime/security/powerauth/app/webflow/configuration/SecurityConfiguration.java (hunk 10 above)
Comment (romanstrobl) [reply powerauth-webflow/src/main/java/io/getlime/security/powerauth/app/webflow/configuration/SecurityConfiguration.java:116]: Sure, why not.

**Pull request merged**